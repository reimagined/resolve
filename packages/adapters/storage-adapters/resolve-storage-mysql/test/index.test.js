import {
  result as mockResult,
  connection as mockConnection
} from 'mysql2/promise'
import createStorageAdapter from '../src/index'

describe('resolve-storage-mysql', () => {
  let storageAdapter = null

  beforeEach(() => {
    storageAdapter = createStorageAdapter({
      host: 'host',
      port: 'port',
      user: 'user',
      password: 'password',
      database: 'database',
      tableName: 'tableName'
    })
  })

  afterEach(async () => {
    await storageAdapter.dispose()

    mockResult.length = 0

    jest.resetAllMocks()
  })

  test('"saveEvent" should save an event with empty payload', async () => {
    await storageAdapter.saveEvent({
      type: 'eventType',
      aggregateId: 'aggregateId',
      aggregateVersion: 1,
      timestamp: 1
    })

    expect(mockConnection.query.mock.calls).toMatchSnapshot()

    expect(mockResult).toMatchSnapshot()
  })

  test('"saveEvent" should save an event', async () => {
    await storageAdapter.saveEvent({
      type: 'eventType',
      aggregateId: 'aggregateId',
      aggregateVersion: 1,
      timestamp: 1,
      payload: { index: 1 }
    })

    expect(mockConnection.query.mock.calls).toMatchSnapshot()

    expect(mockResult).toMatchSnapshot()
  })

  test('"loadEvents" should load events', async () => {
    const loadFilters = [
      {},
      { maxEventsByTimeframe: 1 },
      { eventTypes: ['eventType'] },
      { aggregateIds: ['aggregateId'] },
      { startTime: 1, finishTime: 3 }
    ]

    for (const loadFilter of loadFilters) {
      mockConnection.query.mockReturnValueOnce([[], []])

      await storageAdapter.loadEvents(loadFilter, () => {})
    }

    expect(mockConnection.query.mock.calls).toMatchSnapshot()

    expect(mockResult).toMatchSnapshot()
  })

  test('"getLatestEvent" should get the latest event', async () => {
    const loadFilters = [
      {},
      { eventTypes: ['eventType'] },
      { aggregateIds: ['aggregateId'] },
      { startTime: 1, finishTime: 3 }
    ]

    for (const loadFilter of loadFilters) {
      mockConnection.query.mockReturnValueOnce([[], []])

      await storageAdapter.getLatestEvent(loadFilter)
    }

    expect(mockConnection.query.mock.calls).toMatchSnapshot()

    expect(mockResult).toMatchSnapshot()
  })

  test('"export" should return eventStream', async () => {
    mockConnection.query.mockReturnValueOnce([[], []])

    const eventStream = storageAdapter.export()

    eventStream.on('data', event => {
      expect(JSON.parse(event)).toHaveProperty('payload')
      expect(JSON.parse(event)).toHaveProperty('aggregateVersion')
    })

    await new Promise(done => eventStream.on('end', done))

    expect(mockConnection.query.mock.calls).toMatchSnapshot()

    expect(mockResult).toMatchSnapshot()
  })
})
