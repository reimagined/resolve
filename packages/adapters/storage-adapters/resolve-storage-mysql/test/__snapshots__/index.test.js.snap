// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolve-storage-mysql "export" should return eventStream 1`] = `
Array [
  Array [
    "SELECT * FROM [[[ESCAPEID(tableName)]]]
    ORDER BY [[[ESCAPEID(timestamp)]]] ASC,
    [[[ESCAPEID(aggregateVersion)]]] ASC
    LIMIT 0, 200",
  ],
]
`;

exports[`resolve-storage-mysql "export" should return eventStream 2`] = `
Array [
  "createConnection",
  Array [
    Object {
      "database": "database",
      "host": "host",
      "multipleStatements": true,
      "password": "password",
      "port": "port",
      "user": "user",
    },
  ],
]
`;

exports[`resolve-storage-mysql "getLatestEvent" should get the latest event 1`] = `
Array [
  Array [
    "SELECT * FROM [[[ESCAPEID(tableName)]]] 
    ORDER BY \`timestamp\` DESC, \`aggregateVersion\` DESC",
  ],
  Array [
    "SELECT * FROM [[[ESCAPEID(tableName)]]] WHERE \`type\` IN ([[[ESCAPE(eventType)]]])
    ORDER BY \`timestamp\` DESC, \`aggregateVersion\` DESC",
  ],
  Array [
    "SELECT * FROM [[[ESCAPEID(tableName)]]] WHERE \`aggregateId\` IN ([[[ESCAPE(aggregateId)]]])
    ORDER BY \`timestamp\` DESC, \`aggregateVersion\` DESC",
  ],
  Array [
    "SELECT * FROM [[[ESCAPEID(tableName)]]] WHERE \`timestamp\` > 1 AND \`timestamp\` < 3
    ORDER BY \`timestamp\` DESC, \`aggregateVersion\` DESC",
  ],
]
`;

exports[`resolve-storage-mysql "getLatestEvent" should get the latest event 2`] = `
Array [
  "createConnection",
  Array [
    Object {
      "database": "database",
      "host": "host",
      "multipleStatements": true,
      "password": "password",
      "port": "port",
      "user": "user",
    },
  ],
]
`;

exports[`resolve-storage-mysql "loadEvents" should load events 1`] = `
Array [
  Array [
    "SELECT * FROM [[[ESCAPEID(tableName)]]] 
      ORDER BY [[[ESCAPEID(timestamp)]]] ASC,
      [[[ESCAPEID(aggregateVersion)]]] ASC
      LIMIT 0, 50",
  ],
  Array [
    "SELECT * FROM [[[ESCAPEID(tableName)]]] 
      ORDER BY [[[ESCAPEID(timestamp)]]] ASC,
      [[[ESCAPEID(aggregateVersion)]]] ASC
      LIMIT 0, 50",
  ],
  Array [
    "SELECT * FROM [[[ESCAPEID(tableName)]]] WHERE \`type\` IN ([[[ESCAPE(eventType)]]])
      ORDER BY [[[ESCAPEID(timestamp)]]] ASC,
      [[[ESCAPEID(aggregateVersion)]]] ASC
      LIMIT 0, 50",
  ],
  Array [
    "SELECT * FROM [[[ESCAPEID(tableName)]]] WHERE \`aggregateId\` IN ([[[ESCAPE(aggregateId)]]])
      ORDER BY [[[ESCAPEID(timestamp)]]] ASC,
      [[[ESCAPEID(aggregateVersion)]]] ASC
      LIMIT 0, 50",
  ],
  Array [
    "SELECT * FROM [[[ESCAPEID(tableName)]]] WHERE \`timestamp\` > 1 AND \`timestamp\` < 3
      ORDER BY [[[ESCAPEID(timestamp)]]] ASC,
      [[[ESCAPEID(aggregateVersion)]]] ASC
      LIMIT 0, 50",
  ],
]
`;

exports[`resolve-storage-mysql "loadEvents" should load events 2`] = `
Array [
  "createConnection",
  Array [
    Object {
      "database": "database",
      "host": "host",
      "multipleStatements": true,
      "password": "password",
      "port": "port",
      "user": "user",
    },
  ],
]
`;

exports[`resolve-storage-mysql "saveEvent" should save an event 1`] = `
Array [
  Array [
    "START TRANSACTION;
SELECT [[[ESCAPE(OK)]]]
FROM [[[ESCAPEID(information_schema)]]].[[[ESCAPEID(tables)]]]
WHERE (
    SELECT [[[ESCAPE(Event store is frozen)]]] AS [[[ESCAPEID(EventStoreIsFrozen)]]]
  UNION ALL
    SELECT [[[ESCAPE(Event store is frozen)]]] AS [[[ESCAPEID(EventStoreIsFrozen)]]]
    FROM [[[ESCAPEID(information_schema)]]].[[[ESCAPEID(tables)]]] [[[ESCAPEID(IS)]]]
    WHERE [[[ESCAPEID(IS)]]].[[[ESCAPEID(table_schema)]]] = [[[ESCAPE(database)]]]
    AND [[[ESCAPEID(IS)]]].[[[ESCAPEID(table_name)]]] = [[[ESCAPE(tableName-freeze)]]]
) = [[[ESCAPE(OK)]]];
INSERT INTO [[[ESCAPEID(tableName)]]](
  [[[ESCAPEID(timestamp)]]],
  [[[ESCAPEID(aggregateId)]]],
  [[[ESCAPEID(aggregateVersion)]]],
  [[[ESCAPEID(type)]]],
  [[[ESCAPEID(payload)]]]
) VALUES(
 1,
 [[[ESCAPE(aggregateId)]]],
 1,
 [[[ESCAPE(eventType)]]],
  (CAST([[[ESCAPE({\\"index\\":1})]]] AS JSON))
);
COMMIT;",
  ],
]
`;

exports[`resolve-storage-mysql "saveEvent" should save an event 2`] = `
Array [
  "createConnection",
  Array [
    Object {
      "database": "database",
      "host": "host",
      "multipleStatements": true,
      "password": "password",
      "port": "port",
      "user": "user",
    },
  ],
]
`;

exports[`resolve-storage-mysql "saveEvent" should save an event with empty payload 1`] = `
Array [
  Array [
    "START TRANSACTION;
SELECT [[[ESCAPE(OK)]]]
FROM [[[ESCAPEID(information_schema)]]].[[[ESCAPEID(tables)]]]
WHERE (
    SELECT [[[ESCAPE(Event store is frozen)]]] AS [[[ESCAPEID(EventStoreIsFrozen)]]]
  UNION ALL
    SELECT [[[ESCAPE(Event store is frozen)]]] AS [[[ESCAPEID(EventStoreIsFrozen)]]]
    FROM [[[ESCAPEID(information_schema)]]].[[[ESCAPEID(tables)]]] [[[ESCAPEID(IS)]]]
    WHERE [[[ESCAPEID(IS)]]].[[[ESCAPEID(table_schema)]]] = [[[ESCAPE(database)]]]
    AND [[[ESCAPEID(IS)]]].[[[ESCAPEID(table_name)]]] = [[[ESCAPE(tableName-freeze)]]]
) = [[[ESCAPE(OK)]]];
INSERT INTO [[[ESCAPEID(tableName)]]](
  [[[ESCAPEID(timestamp)]]],
  [[[ESCAPEID(aggregateId)]]],
  [[[ESCAPEID(aggregateVersion)]]],
  [[[ESCAPEID(type)]]],
  [[[ESCAPEID(payload)]]]
) VALUES(
 1,
 [[[ESCAPE(aggregateId)]]],
 1,
 [[[ESCAPE(eventType)]]],
  (CAST([[[ESCAPE(null)]]] AS JSON))
);
COMMIT;",
  ],
]
`;

exports[`resolve-storage-mysql "saveEvent" should save an event with empty payload 2`] = `
Array [
  "createConnection",
  Array [
    Object {
      "database": "database",
      "host": "host",
      "multipleStatements": true,
      "password": "password",
      "port": "port",
      "user": "user",
    },
  ],
]
`;
